flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/plan/utils/FlinkRexUtil.scala

```diff
@@ -650,6 +654,75 @@ object FlinkRexUtil {
       rexBuilder,
       converter);
   }
+
+  /**
+   * Return two neighbouring [[Project]] can merge into one [[Project]] or not. If the two
+   * [[Project]] can merge into one, each non-deterministic [[RexNode]] of bottom [[Project]] should
+   * appear at most once in the project list of top [[Project]].
+   */
+  def isMergeable(topProject: Project, bottomProject: Project): Boolean = {
+    val topInputRefCounter: Array[Int] =
+      Array.fill(topProject.getInput.getRowType.getFieldCount)(0)
+
+    mergeable(topInputRefCounter, topProject.getProjects, bottomProject.getProjects)
+  }
+
+  /**
+   * An InputRefCounter that count every inputRef's reference count number, every reference will be
+   * counted, e.g., '$0 + 1' & '$0 + 2' will count 2 instead of 1.
+   * @param deep
+   * @param refCounts
+   */
+  private class InputRefCounter(deep: Boolean, val refCounts: Array[Int])
+    extends RexVisitorImpl[Void](deep: Boolean) {
+    override def visitInputRef(inputRef: RexInputRef): Void = {
+      val index = inputRef.getIndex
+      refCounts(index) += 1
+      null
+    }
+  }
+
+  private def mergeable(
+      topInputRefCounter: Array[Int],
+      topProjects: JList[RexNode],
+      bottomProjects: JList[RexNode]): Boolean = {
+    RexUtil.apply(new InputRefCounter(true, topInputRefCounter), topProjects, null)
+
+    bottomProjects.zipWithIndex
+      .map {
+        case (p, idx) =>
+          if (!RexUtil.isDeterministic(p)) {
+            topInputRefCounter(idx)
+          } else 0
+      }
+      .forall(cnt => cnt <= 1)
+  }
+
+  /**
+   * Return two neighbouring [[Calc]] can merge into one [[Calc]] or not. If the two [[Calc]] can
+   * merge into one, each non-deterministic [[RexNode]] of bottom [[Calc]] should appear at most
+   * once in the project list and filter list of top [[Calc]].
+   */
+  def isMergeable(topCalc: Calc, bottomCalc: Calc): Boolean = {
+    val topProgram = topCalc.getProgram
+    val bottomProgram = bottomCalc.getProgram
+    val topInputRefCounter: Array[Int] =
+      Array.fill(topCalc.getInput.getRowType.getFieldCount)(0)
+
+    val topInputRefs = if (null != topProgram.getCondition) {
+      topProgram.getProjectList.map(topProgram.expandLocalRef) :+ topProgram.expandLocalRef(
+        topProgram.getCondition)
+    } else {
+      topProgram.getProjectList.map(topProgram.expandLocalRef)
+    }.toList
+
+    mergeable(
+      topInputRefCounter,
+      topInputRefs.toList,
+      bottomProgram.getProjectList
+        .map(bottomProgram.expandLocalRef)
+        .toList)
+  }
```

Comment:
Nits: To achieve the goal of scala-free, would it be better to put these utils methods in a Java class?

Reply:
Ok, this encourage me creating a new class `FlinkRelUtil` which is more appropriately than the `FlinkRexUtil` here